File: sdks/fabric/go-sdk/asset-manager/assetManager.go
/*
Copyright 2020 IBM All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package assetmanager

import (
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"fmt"
	"time"

	"github.com/hyperledger-labs/weaver-dlt-interoperability/common/protos-go/common"
	log "github.com/sirupsen/logrus"

	"github.com/golang/protobuf/proto"
)

type GatewayContract interface {
	SubmitTransaction(string, ...string) ([]byte, error)
	EvaluateTransaction(string, ...string) ([]byte, error)
}

// helper functions to log and return errors
func logThenErrorf(format string, args ...interface{}) error {
	errorMsg := fmt.Sprintf(format, args...)
	log.Error(errorMsg)
	return errors.New(errorMsg)
}

// Create an asset exchange agreement structure
func createAssetExchangeAgreementSerializedBase64(assetType string, assetId string, recipientECertBase64 string, lockerECertBase64 string) (string, error) {
	assetAgreement := &common.AssetExchangeAgreement{
		AssetType: assetType,
		Id:        assetId,
		Recipient: recipientECertBase64,
		Locker:    lockerECertBase64,
	}
	assetAgreementBytes, err := proto.Marshal(assetAgreement)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	return base64.StdEncoding.EncodeToString(assetAgreementBytes), nil
}

// Create a fungible asset exchange agreement structure
func createFungibleAssetExchangeAgreementSerializedBase64(assetType string, numUnits uint64, recipientECertBase64 string, lockerECertBase64 string) (string, error) {
	assetAgreement := &common.FungibleAssetExchangeAgreement{
		AssetType: assetType,
		NumUnits:  numUnits,
		Recipient: recipientECertBase64,
		Locker:    lockerECertBase64,
	}
	assetAgreementBytes, err := proto.Marshal(assetAgreement)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	return base64.StdEncoding.EncodeToString(assetAgreementBytes), nil
}

// Create an asset lock structure
func createAssetLockInfoSerializedBase64(hashBase64 string, expiryTimeSecs uint64) (string, error) {
	lockInfoHTLC := &common.AssetLockHTLC{
		HashBase64:     []byte(hashBase64),
		ExpiryTimeSecs: expiryTimeSecs,
		TimeSpec:       common.TimeSpec_EPOCH,
	}
	lockInfoHTLCBytes, err := proto.Marshal(lockInfoHTLC)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	lockInfo := &common.AssetLock{
		LockMechanism: common.LockMechanism_HTLC,
		LockInfo:      lockInfoHTLCBytes,
	}
	lockInfoBytes, err := proto.Marshal(lockInfo)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	return base64.StdEncoding.EncodeToString(lockInfoBytes), nil
}

// Create an asset claim structure
func createAssetClaimInfoSerializedBase64(hashPreimageBase64 string) (string, error) {
	claimInfoHTLC := &common.AssetClaimHTLC{
		HashPreimageBase64: []byte(hashPreimageBase64),
	}
	claimInfoHTLCBytes, err := proto.Marshal(claimInfoHTLC)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	claimInfo := &common.AssetClaim{
		LockMechanism: common.LockMechanism_HTLC,
		ClaimInfo:     claimInfoHTLCBytes,
	}
	claimInfoBytes, err := proto.Marshal(claimInfo)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	return base64.StdEncoding.EncodeToString(claimInfoBytes), nil
}

// function to generate a "SHA256" hash in base64 format for a given preimage
func GenerateSHA256HashInBase64Form(hashPreimage string) string {
	hasher := sha256.New()
	hasher.Write([]byte(hashPreimage))
	shaHash := hasher.Sum(nil)
	shaHashBase64 := base64.StdEncoding.EncodeToString(shaHash)

	return shaHashBase64
}

func CreateHTLC(contract GatewayContract, assetType string, assetId string, recipientECertBase64 string,
	hashBase64 string, expiryTimeSecs uint64) (string, error) {
	if contract == nil {
		return "", logThenErrorf("contract handle not supplied")
	}
	if assetType == "" {
		return "", logThenErrorf("asset type not supplied")
	}
	if assetId == "" {
		return "", logThenErrorf("asset id not supplied")
	}
	if recipientECertBase64 == "" {
		return "", logThenErrorf("recipientECertBase64 id not supplied")
	}
	if hashBase64 == "" {
		return "", logThenErrorf("hashBase64 is not supplied")
	}
	currentTimeSecs := uint64(time.Now().Unix())
	if expiryTimeSecs <= currentTimeSecs {
		return "", logThenErrorf("supplied expirty time in the past")
	}

	assetExchangeAgreementStr, err := createAssetExchangeAgreementSerializedBase64(assetType, assetId, recipientECertBase64, "")
	if err != nil {
		return "", logThenErrorf(err.Error())
	}
	lockInfoStr, err := createAssetLockInfoSerializedBase64(hashBase64, expiryTimeSecs)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	// Normal invoke function
	result, err := contract.SubmitTransaction("LockAsset", assetExchangeAgreementStr, lockInfoStr)
	if err != nil {
		return "", logThenErrorf("error in contract.SubmitTransaction LockAsset: %+v", err.Error())
	}

	return string(result), nil
}

func CreateFungibleHTLC(contract GatewayContract, assetType string, numUnits uint64, recipientECertBase64 string,
	hashBase64 string, expiryTimeSecs uint64) (string, error) {
	if contract == nil {
		return "", logThenErrorf("contract handle not supplied")
	}
	if assetType == "" {
		return "", logThenErrorf("asset type not supplied")
	}
	if numUnits <= 0 {
		return "", logThenErrorf("asset count must be a positive number")
	}
	if recipientECertBase64 == "" {
		return "", logThenErrorf("recipientECertBase64 id not supplied")
	}
	if hashBase64 == "" {
		return "", logThenErrorf("hashBase64 is not supplied")
	}
	currentTimeSecs := uint64(time.Now().Unix())
	if expiryTimeSecs <= currentTimeSecs {
		return "", logThenErrorf("supplied expirty time in the past")
	}

	assetExchangeAgreementStr, err := createFungibleAssetExchangeAgreementSerializedBase64(assetType, numUnits, recipientECertBase64, "")
	if err != nil {
		return "", logThenErrorf(err.Error())
	}
	lockInfoStr, err := createAssetLockInfoSerializedBase64(hashBase64, expiryTimeSecs)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	// Normal invoke function
	result, err := contract.SubmitTransaction("LockFungibleAsset", assetExchangeAgreementStr, lockInfoStr)
	if err != nil {
		return "", logThenErrorf("error in contract.SubmitTransaction LockFungibleAsset: %+v", err.Error())
	}

	return string(result), nil
}

func IsAssetLockedInHTLC(contract GatewayContract, assetType string, assetId string, recipientECertBase64 string, lockerECertBase64 string) (string, error) {

	if contract == nil {
		return "", logThenErrorf("contract handle not supplied")
	}
	if assetType == "" {
		return "", logThenErrorf("asset type not supplied")
	}
	if assetId == "" {
		return "", logThenErrorf("asset id not supplied")
	}
	if recipientECertBase64 == "" {
		return "", logThenErrorf("recipientECertBase64 id not supplied")
	}
	if lockerECertBase64 == "" {
		return "", logThenErrorf("lockerECertBase64 id not supplied")
	}

	assetExchangeAgreementStr, err := createAssetExchangeAgreementSerializedBase64(assetType, assetId, recipientECertBase64, lockerECertBase64)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	// Normal invoke function
	result, err := contract.EvaluateTransaction("IsAssetLocked", assetExchangeAgreementStr)
	if err != nil {
		return "", logThenErrorf("error in contract.EvaluateTransaction IsAssetLocked: %+v", err.Error())
	}

	return string(result), nil
}

func IsFungibleAssetLockedInHTLC(contract GatewayContract, contractId string) (string, error) {

	if contract == nil {
		return "", logThenErrorf("contract handle not supplied")
	}
	if contractId == "" {
		return "", logThenErrorf("contractId not supplied")
	}

	// Normal invoke function
	result, err := contract.EvaluateTransaction("IsFungibleAssetLocked", contractId)
	if err != nil {
		return "", logThenErrorf("error in contract.EvaluateTransaction IsFungibleAssetLocked: %+v", err.Error())
	}

	return string(result), nil
}

func IsAssetLockedInHTLCqueryUsingContractId(contract GatewayContract, contractId string) (string, error) {

	if contract == nil {
		return "", logThenErrorf("contract handle not supplied")
	}
	if contractId == "" {
		return "", logThenErrorf("contractId not supplied")
	}

	// Normal invoke function
	result, err := contract.EvaluateTransaction("IsAssetLockedQueryUsingContractId", contractId)
	if err != nil {
		return "", logThenErrorf("error in contract.EvaluateTransaction IsAssetLockedQueryUsingContractId: %+v", err.Error())
	}

	return string(result), nil
}

func ClaimAssetInHTLC(contract GatewayContract, assetType string, assetId string, lockerECertBase64 string, hashPreimageBase64 string) (string, error) {
	if contract == nil {
		return "", logThenErrorf("contract handle not supplied")
	}
	if assetType == "" {
		return "", logThenErrorf("asset type not supplied")
	}
	if assetId == "" {
		return "", logThenErrorf("asset id not supplied")
	}
	if lockerECertBase64 == "" {
		return "", logThenErrorf("lockerECertBase64 id not supplied")
	}
	if hashPreimageBase64 == "" {
		return "", logThenErrorf("hashPreimageBase64 is not supplied")
	}

	claimInfoStr, err := createAssetClaimInfoSerializedBase64(hashPreimageBase64)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	assetExchangeAgreementStr, err := createAssetExchangeAgreementSerializedBase64(assetType, assetId, "", lockerECertBase64)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	// Normal invoke function
	result, err := contract.SubmitTransaction("ClaimAsset", assetExchangeAgreementStr, claimInfoStr)
	if err != nil {
		return "", logThenErrorf("error in contract.SubmitTransaction ClaimAsset: %+v", err.Error())
	}

	return string(result), nil
}

func ClaimFungibleAssetInHTLC(contract GatewayContract, contractId string, hashPreimageBase64 string) (string, error) {
	if contract == nil {
		return "", logThenErrorf("contract handle not supplied")
	}
	if contractId == "" {
		return "", logThenErrorf("contractId not supplied")
	}
	if hashPreimageBase64 == "" {
		return "", logThenErrorf("hashPreimageBase64 is not supplied")
	}

	claimInfoStr, err := createAssetClaimInfoSerializedBase64(hashPreimageBase64)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	// Normal invoke function
	result, err := contract.SubmitTransaction("ClaimFungibleAsset", contractId, claimInfoStr)
	if err != nil {
		return "", logThenErrorf("error in contract.SubmitTransaction ClaimFungibleAsset: %+v", err.Error())
	}

	return string(result), nil
}

func ClaimAssetInHTLCusingContractId(contract GatewayContract, contractId string, hashPreimageBase64 string) (string, error) {
	if contract == nil {
		return "", logThenErrorf("contract handle not supplied")
	}
	if contractId == "" {
		return "", logThenErrorf("contractId not supplied")
	}
	if hashPreimageBase64 == "" {
		return "", logThenErrorf("hashPreimageBase64 is not supplied")
	}

	claimInfoStr, err := createAssetClaimInfoSerializedBase64(hashPreimageBase64)
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	// Normal invoke function
	result, err := contract.SubmitTransaction("ClaimAssetUsingContractId", contractId, claimInfoStr)
	if err != nil {
		return "", logThenErrorf("error in contract.SubmitTransaction ClaimAssetUsingContractId: %+v", err.Error())
	}

	return string(result), nil
}

func ReclaimAssetInHTLC(contract GatewayContract, assetType string, assetId string, recipientECertBase64 string) (string, error) {
	if contract == nil {
		return "", logThenErrorf("contract handle not supplied")
	}
	if assetType == "" {
		return "", logThenErrorf("asset type not supplied")
	}
	if assetId == "" {
		return "", logThenErrorf("asset id not supplied")
	}
	if recipientECertBase64 == "" {
		return "", logThenErrorf("recipientECertBase64 id not supplied")
	}

	assetExchangeAgreementStr, err := createAssetExchangeAgreementSerializedBase64(assetType, assetId, recipientECertBase64, "")
	if err != nil {
		return "", logThenErrorf(err.Error())
	}

	// Normal invoke function
	result, err := contract.SubmitTransaction("UnlockAsset", assetExchangeAgreementStr)
	if err != nil {
		return "", logThenErrorf("error in contract.SubmitTransaction UnlockAsset: %+v", err.Error())
	}

	return string(result), nil
}

func ReclaimFungibleAssetInHTLC(contract GatewayContract, contractId string) (string, error) {
	if contract == nil {
		return "", logThenErrorf("contract handle not supplied")
	}
	if contractId == "" {
		return "", logThenErrorf("contractId not supplied")
	}

	// Normal invoke function
	result, err := contract.SubmitTransaction("UnlockFungibleAsset", contractId)
	if err != nil {
		return "", logThenErrorf("error in contract.SubmitTransaction UnlockFungibleAsset: %+v", err.Error())
	}

	return string(result), nil
}

func ReclaimAssetInHTLCusingContractId(contract GatewayContract, contractId string) (string, error) {
	if contract == nil {
		return "", logThenErrorf("contract handle not supplied")
	}
	if contractId == "" {
		return "", logThenErrorf("contractId not supplied")
	}

	// Normal invoke function
	result, err := contract.SubmitTransaction("UnlockAssetUsingContractId", contractId)
	if err != nil {
		return "", logThenErrorf("error in contract.SubmitTransaction UnlockAssetUsingContractId: %+v", err.Error())
	}

	return string(result), nil
}

File: weaver/sdks/besu/interoperation-node-sdk/src/AssetManager.ts
// * SPDX-License-Identifier: Apache-2.0
// */

/**
* This file provides helper functions for interoperability operations.
**/
/** End file docs */
import * as assetLocksPb from "@hyperledger-labs/weaver-protos-js/common/asset_locks_pb";
import * as Web3 from 'web3-utils';
import { Hash, SHA256 } from "./HashFunctions";

// Create an asset exchange agreement structure
function createAssetExchangeAgreementSerialized(assetType: string, assetID: string, locker: string, recipient: string) {
    const assetExchangeAgreement = new assetLocksPb.AssetExchangeAgreement();
    assetExchangeAgreement.setAssettype(assetType);
    assetExchangeAgreement.setId(assetID);
    assetExchangeAgreement.setRecipient(recipient);
    assetExchangeAgreement.setLocker(locker);
    return Buffer.from(assetExchangeAgreement.serializeBinary())
}

function createFungibleAssetExchangeAgreementSerialized(assetType: string, numUnits: number, locker: string, recipient: string) {
    const assetExchangeAgreement = new assetLocksPb.FungibleAssetExchangeAgreement();
    assetExchangeAgreement.setAssettype(assetType);
    assetExchangeAgreement.setNumunits(numUnits);
    assetExchangeAgreement.setRecipient(recipient);
    assetExchangeAgreement.setLocker(locker);
    return Buffer.from(assetExchangeAgreement.serializeBinary())
}


function createHybridAssetExchangeAgreementSerialized(assetType: string, assetID: string, assetData: string, numUnits: number, recipient: string) {
    const assetExchangeAgreement = new assetLocksPb.HybridAssetExchangeAgreement();
    assetExchangeAgreement.setAssettype(assetType);
    assetExchangeAgreement.setId(assetID);
    assetExchangeAgreement.setRecipient(recipient);
    assetExchangeAgreement.setAssetdata(Buffer.from(assetData, "hex"));
    assetExchangeAgreement.setNumunits(numUnits);
    return Buffer.from(assetExchangeAgreement.serializeBinary())
}

// Create an asset lock structure
function createAssetLockInfoSerialized(hash: Hash, expiryTimeSecs: number) {
    const lockInfoHTLC = new assetLocksPb.AssetLockHTLC();
    lockInfoHTLC.setHashmechanism(hash.HASH_MECHANISM);
    lockInfoHTLC.setHashbase64(Buffer.from(hash.getSerializedHashBase64(), 'base64'));
    lockInfoHTLC.setExpirytimesecs(expiryTimeSecs);
    lockInfoHTLC.setTimespec(assetLocksPb.TimeSpec.EPOCH)
    return Buffer.from(lockInfoHTLC.serializeBinary())
}

// Create an asset claim structure
function createAssetClaimInfoSerialized(hash: Hash) {
    const claimInfoHTLC = new assetLocksPb.AssetClaimHTLC();
    claimInfoHTLC.setHashmechanism(hash.HASH_MECHANISM);
    claimInfoHTLC.setHashpreimagebase64(Buffer.from(hash.getPreimage()));
    const claimInfoHTLCSerialized = claimInfoHTLC.serializeBinary();
    const claimInfo = new assetLocksPb.AssetClaim();
    claimInfo.setLockmechanism(assetLocksPb.LockMechanism.HTLC);
    claimInfo.setClaiminfo(claimInfoHTLCSerialized);
    return Buffer.from(claimInfo.serializeBinary())
}

const createHTLC = async (
    assetManagerContract: any,
    appContract: any,
    assetType: string,
    assetID: string,
    lockerAddress: string,
    recipientAddress: string,
    expiryTimeSecs: number,
    hash: Hash
): Promise<{ result: any, hash: Hash }> => {

    if (!assetManagerContract || !appContract) {
        console.log("Contract handle not supplied");
        return { result: false, hash: null };
    }
    if (!recipientAddress) {
        console.log(`Recipient address not supplied ${recipientAddress}`);
        return { result: false, hash: null };
    }
    const currTimeSecs = Math.floor(Date.now() / 1000);   // Convert epoch milliseconds to seconds
    if (expiryTimeSecs <= currTimeSecs) {
        console.log("Supplied expiry time invalid or in the past: %s; current time: %s", new Date(expiryTimeSecs).toISOString(), new Date(currTimeSecs).toISOString());
        return { result: false, hash: null };
    }
    
    if (hash == null) {
        hash = new SHA256()
    }
    if (hash.hash64 == null) {
        hash.generateRandomPreimage(22);
    }
    
    var agreement = createAssetExchangeAgreementSerialized(assetType, assetID, "", recipientAddress.slice(2));

    if (!agreement) {
        console.log("Error creating protobuf asset exchange agreement");
        return { result: false, hash: null };
    }
    const lockInfo = createAssetLockInfoSerialized(hash, expiryTimeSecs);
    // Normal invoke function
    let lockStatus = await assetManagerContract.lockNonFungibleAsset(
        appContract.address,
        agreement,
        lockInfo,
        {
            from: lockerAddress,
        }
    ).catch(function (e: any) {
        console.log(e);
        console.log("lockNonFungibleAsset threw an error");
        lockStatus = { result: false, hash: null }
    })
    return { result: lockStatus, hash: hash };
};

const createFungibleHTLC = async (
    assetManagerContract: any,
    appContract: any,
    assetType: string,
    assetAmount: number,
    lockerAddress: string,
    recipientAddress: string,
    expiryTimeSecs: number,
    hash: Hash
): Promise<{ result: any, hash: Hash }> => {

    if (!assetManagerContract || !appContract) {
        console.log("Contract handle not supplied");
        return { result: false, hash: null };
    }
    if (!recipientAddress) {
        console.log(`Recipient address not supplied ${recipientAddress}`);
        return { result: false, hash: null };
    }
    const currTimeSecs = Math.floor(Date.now() / 1000);   // Convert epoch milliseconds to seconds
    if (expiryTimeSecs <= currTimeSecs) {
        console.log("Supplied expiry time invalid or in the past: %s; current time: %s", new Date(expiryTimeSecs).toISOString(), new Date(currTimeSecs).toISOString());
        return { result: false, hash: null };
    }
    
    if (hash == null) {
        hash = new SHA256()
    }
    if (hash.hash64 == null) {
        hash.generateRandomPreimage(22);
    }
    
    var agreement = createFungibleAssetExchangeAgreementSerialized(assetType, assetAmount, "", recipientAddress.slice(2));

    if (!agreement) {
        console.log("Error creating protobuf fungible asset exchange agreement");
        return { result: false, hash: null };
    }
    const lockInfo = createAssetLockInfoSerialized(hash, expiryTimeSecs);
    // Normal invoke function
    let lockStatus = await assetManagerContract.lockFungibleAsset(
        appContract.address,
        agreement,
        lockInfo,
        {
            from: lockerAddress,
        }
    ).catch(function (e: any) {
        console.log(e);
        console.log("lockFungibleAsset threw an error");
        lockStatus = { result: false, hash: null }
    })
    return { result: lockStatus, hash: hash };
};

const createHybridHTLC = async (
    assetManagerContract: any,
    appContract: any,
    assetType: string,
    assetID: string,
    assetData: string,
    assetAmount: number,
    lockerAddress: string,
    recipientAddress: string,
    expiryTimeSecs: number,
    hash: Hash
): Promise<{ result: any, hash: Hash }> => {

    if (!assetManagerContract || !appContract) {
        console.log("Contract handle not supplied");
        return { result: false, hash: null };
    }
    if (!recipientAddress) {
        console.log(`Recipient address not supplied ${recipientAddress}`);
        return { result: false, hash: null };
    }
    const currTimeSecs = Math.floor(Date.now() / 1000);   // Convert epoch milliseconds to seconds
    if (expiryTimeSecs <= currTimeSecs) {
        console.log("Supplied expiry time invalid or in the past: %s; current time: %s", new Date(expiryTimeSecs).toISOString(), new Date(currTimeSecs).toISOString());
        return { result: false, hash: null };
    }
    
    if (hash == null) {
        hash = new SHA256()
    }
    if (hash.hash64 == null) {
        hash.generateRandomPreimage(22);
    }
    
    var agreement = createHybridAssetExchangeAgreementSerialized(assetType, String(assetID), Web3.utf8ToHex(String(assetData)), assetAmount, recipientAddress.slice(2));
    if (!agreement) {
        console.log("Error creating protobuf hybrid asset exchange agreement");
        return { result: false, hash: null };
    }
    const lockInfo = createAssetLockInfoSerialized(hash, expiryTimeSecs);
    // Normal invoke function
    let lockStatus = await assetManagerContract.lockHybridAsset(
        appContract.address,
        agreement,
        lockInfo,
        {
            from: lockerAddress,
        }
    ).catch(function (e: any) {
        console.log(e);
        console.log("lockHybridAsset threw an error");
        lockStatus = { result: false, hash: null }
    })
    
    return { result: lockStatus, hash: hash };
};

/**
 * Latter step of a Hashed Time Lock Contract
 * - Claim a unique asset instance using a hash preimage
 **/
const claimAssetInHTLC = async (
    assetManagerContract: any,
    contractId: string,
    lockerAddress: string,
    hash: Hash,
): Promise<any> => {

    if (!assetManagerContract) {
        console.log("Contract Address not supplied");
        return false;
    }

    const claimInfoStr = createAssetClaimInfoSerialized(hash);
    
    var claimStatus = await assetManagerContract.claimAsset(contractId, claimInfoStr, { from: lockerAddress }).catch(function (e: any) {
        console.log(e)
        console.log("claimAsset threw an error");
        claimStatus = false
    })

    return claimStatus;
};

const isAssetLockedInHTLC = async (
    assetManagerContract: any,
    contractId: string,
): Promise<any> => {

    if (!assetManagerContract) {
        console.log("Contract not supplied");
        return false;
    }

    // Normal invoke function
    var lockStatus = await assetManagerContract.isAssetLocked(contractId).catch(function (e: any) {
        console.log(e)
        console.log("isAssetLock threw an error");
        lockStatus = false
    })

    return lockStatus;
};


const reclaimAssetInHTLC = async (
    interopContract: any,
    contractId: string,
    locker: string,
): Promise<void> => {
    var unlockStatus = await interopContract.unlockAsset(contractId, {
        from: locker
    }).catch(function (e: any) {
        console.log(e)
        console.log("unlockAsset threw an error");
        unlockStatus = false
    })

    return unlockStatus
}

export {
    createAssetExchangeAgreementSerialized,
    createAssetClaimInfoSerialized,
    createHTLC,
    createFungibleHTLC,
    createHybridHTLC,
    claimAssetInHTLC,
    reclaimAssetInHTLC,
    isAssetLockedInHTLC
};
